File: Client/src/main/java/com/rspsi/editor/model/ObjectCategory.java
package com.rspsi.editor.model;

/** Neutral semantic layer for an OSRS location shape. */
public enum ObjectCategory {
    WALL(0, "Wall"),
    WALL_DECOR(1, "Wall decor"),
    GROUND(2, "Game object"),
    GROUND_DECOR(3, "Ground decor"),
    UNKNOWN(-1, "Unknown");

    private final int layerId;
    private final String displayName;

    ObjectCategory(int layerId, String displayName) {
        this.layerId = layerId;
        this.displayName = displayName;
    }

    public int layerId() {
        return layerId;
    }

    public String displayName() {
        return displayName;
    }

    public boolean isKnown() {
        return this != UNKNOWN;
    }

    /** Maps the canonical OSRS location-shape range to its scene layer. */
    public static ObjectCategory fromType(int type) {
        return OsrsLocShape.fromId(type).map(OsrsLocShape::category).orElse(UNKNOWN);
    }
}

File: Client/src/main/kotlin/com/rspsi/editor/model/ObjectCategory.kt
package com.rspsi.editor.model

/**
 * Neutral semantic scene layer for an OSRS location shape.
 *
 * The editor reasons about these categories independently of renderer-specific object
 * classes so inspectors, tools, and serialization can share one canonical mapping.
 */
enum class ObjectCategory(
    private val layerIdValue: Int,
    private val displayNameValue: String,
) {
    WALL(0, "Wall"),
    WALL_DECOR(1, "Wall decor"),
    GROUND(2, "Game object"),
    GROUND_DECOR(3, "Ground decor"),
    UNKNOWN(-1, "Unknown");

    fun layerId(): Int = layerIdValue

    fun displayName(): String = displayNameValue

    fun isKnown(): Boolean = this != UNKNOWN

    companion object {
        /** Maps the canonical OSRS location-shape range to its neutral scene layer. */
        @JvmStatic
        fun fromType(type: Int): ObjectCategory =
            OsrsLocShape.fromId(type)
                .map { it.category() }
                .orElse(UNKNOWN)
    }
}
